### PR TITLE
[CARBONDATA-2755][Complex DataType Enhancements] Compaction Complex Types (STRUCT AND ARRAY)

### PR DESCRIPTION
**Pull request by dhatchayani**
**Cherry-Picked from PR  https://github.com/apache/carbondata/pull/2516**

(1) Enabling Compaction with Complex DataTypes.
(2) Major Minor compaction will run over complex dataTypes.

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [x] Testing done
        UT Added
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1643/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9903/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1854/

**Comment by kumarvishal09** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1888/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9936/

File: processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java [hunk 1]
```diff
@@ -337,6 +342,25 @@ public static void addColumnCardinalityToMap(Map<String, Integer> columnCardinal
         .toPrimitive(updatedCardinalityList.toArray(new Integer[updatedCardinalityList.size()]));
   }
 
+  private static void fillColumnSchemaListForComplexDims(List<CarbonDimension> carbonDimensionsList,
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java]: Can you add comment what is happening in this method?

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java [hunk 2]
```diff
@@ -371,9 +374,25 @@ public static CarbonFactDataHandlerModel getCarbonFactDataHandlerModel(CarbonLoa
         .getFormattedCardinality(segmentProperties.getDimColumnsCardinality(), wrapperColumnSchema);
     carbonFactDataHandlerModel.setColCardinality(formattedCardinality);
     //TO-DO Need to handle complex types here .
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: Remove it

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java [hunk 3]
```diff
@@ -407,6 +426,81 @@ public static CarbonFactDataHandlerModel getCarbonFactDataHandlerModel(CarbonLoa
...
+  private static Map<Integer, GenericDataType> convertComplexDimensionToGenericDataType(
+      List<CarbonDimension> complexDimensions, int dimensionCount, String isNullFormat,
+      boolean isEmptyBadRecords) {
+    Map<Integer, GenericDataType> complexIndexMap =
+        new HashMap<Integer, GenericDataType>(complexDimensions.size());
+
+    for (CarbonDimension carbonDimension : complexDimensions) {
+
+      if (carbonDimension.isComplex()) {
+        GenericDataType g;
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: Pls give some proper name

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java [hunk 4]
```diff
@@ -407,6 +426,81 @@ public static CarbonFactDataHandlerModel getCarbonFactDataHandlerModel(CarbonLoa
...
+        if (carbonDimension.getColumnSchema().getDataType().getName().equalsIgnoreCase("ARRAY")) {
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: Please check the utility to get the complex type

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java [hunk 5]
```diff
@@ -407,6 +426,81 @@ public static CarbonFactDataHandlerModel getCarbonFactDataHandlerModel(CarbonLoa
...
+        if (carbonDimension.getNumberOfChild() > 0) {
+          addChildrenForComplex(carbonDimension.getListOfChildDimensions(), arrayGeneric,
+              isNullFormat, isEmptyBadRecord);
+        }
+        genericDataType.addChildren(arrayGeneric);
+      } else if (carbonDimension.getColumnSchema().getDataType().getName()
+          .equalsIgnoreCase("STRUCT")) {
+        GenericDataType structGeneric = new StructDataType(carbonDimension.getColName(),
+            carbonDimension.getColName()
+                .substring(0, carbonDimension.getColName().lastIndexOf(".")),
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: Pls extract to top and reuse it

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java [hunk 6]
```diff
@@ -407,6 +426,81 @@ public static CarbonFactDataHandlerModel getCarbonFactDataHandlerModel(CarbonLoa
...
+        genericDataType.addChildren(structGeneric);
+      } else {
+        // Primitive Data Type
+        genericDataType.addChildren(
+            new PrimitiveDataType(carbonDimension.getColumnSchema().getColumnName(),
+                carbonDimension.getDataType(), carbonDimension.getColName()
+                .substring(0, carbonDimension.getColName().lastIndexOf(".")),
+                carbonDimension.getColumnId(),
+                carbonDimension.getColumnSchema().hasEncoding(Encoding.DICTIONARY), isNullFormat,
+                isEmptyBadRecord));
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: Please check and remove `isEmptyBadRecord` from it if not used

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java [hunk 7]
```diff
@@ -371,9 +374,25 @@ public static CarbonFactDataHandlerModel getCarbonFactDataHandlerModel(CarbonLoa
...
-        new HashMap<Integer, GenericDataType>(segmentProperties.getComplexDimensions().size());
-    carbonFactDataHandlerModel.setComplexIndexMap(complexIndexMap);
+
+    int simpleDimensionCount = -1;
+    if (segmentProperties.getDimensions().size() == 0) {
+      simpleDimensionCount = 0;
+    } else {
+      simpleDimensionCount = segmentProperties.getDimensions().size() - segmentProperties
+          .getNumberOfNoDictionaryDimension() - segmentProperties.getComplexDimensions().size();
+    }
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: Please move down this code to `convertComplexDimensionToGenericDataType`

**Comment by ravipesala** [comment]
This PR supports compaction only for STRUCT and ARRAY. Please raise another jira and PR to support MAP type as well.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1685/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9945/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1897/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1693/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1696/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1906/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9956/

**Comment by dhatchayani** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1699/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9959/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1909/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1731/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1942/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9991/

**Pull request closed without merging**